### PR TITLE
Add device wrapper

### PR DESCRIPTION
**Pull request by Nugine**

### CODE REVIEW

**Comment by Nugine** [comment]
Request review @GTwhy @rogercloud

File: src/context.rs [hunk 1]
```diff
@@ -47,54 +46,23 @@ impl Context {
...
-                    dev_name_inner.eq(name)
-                })
-                .ok_or(io::ErrorKind::NotFound)?
-        } else {
-            dev_list.get(0).ok_or(io::ErrorKind::NotFound)?
-        };
+        .ok_or(io::ErrorKind::NotFound)?;
+
         // SAFETY: ffi
         // TODO: check safety
```
Comment (GTwhy) [inline src/context.rs:61]: Do we need more checks here?
For ibv APIs, if we correctly pass a value returned by another ibv API, its safety depends on the ibv API itself.
So I think only `SAFETY: ffi` is ok here, what's your opinion?

**Review by GTwhy — COMMENTED** [review]
LGTM.
A pr with educational significance :-)

File: src/context.rs (hunk 1 above)
Comment (Nugine) [reply src/context.rs]: Added some explanations in the latest commit

File: src/device.rs [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+        // SAFETY: ffi
+        unsafe { CStr::from_ptr(ibv_get_device_name(self.ffi_ptr())) }
+    }
+
+    /// Returns kernel device name
+    #[inline]
+    #[must_use]
+    pub fn name(&self) -> &str {
+        #[allow(clippy::expect_used)]
+        self.c_name().to_str().expect("non-utf8 device name")
```
Comment (rogercloud) [inline src/device.rs:132]: Should we panic? Maybe return option is a better choise. Sometimes users just want to ignore these names.

Comment (rogercloud) [inline src/device.rs:132]: Should we panic? Maybe return option is a better choise. Sometimes users just want to ignore these names.

Comment (rogercloud) [reply src/device.rs:132]: If panic is necessary please doc it in the comment.

File: src/device.rs [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+
+impl fmt::Debug for Guid {
+    #[inline]
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Guid({:x})", self)
+    }
+}
+
+/// Encodes `src` to a hex string and process it
+fn be64_to_hex<R>(src: __be64, case: hex_simd::AsciiCase, f: impl FnOnce(&str) -> R) -> R {
```
Comment (rogercloud) [inline src/device.rs:174]: Accept src as Guid is better as we can use `as_bytes` method here.

Comment (rogercloud) [inline src/device.rs:174]: Accept src as Guid is better as we can use `as_bytes` method here.

File: src/device.rs (hunk 2 above)
Comment (Nugine) [reply src/device.rs:135]: In most cases, the device name is an ascii string. A panicking method (`guid.name()`) is more convenient than handling a c string with unknown encoding (`guid.c_name()`)
Added panics documentation in the latest force-update.

File: Cargo.toml [hunk 4]
```diff
@@ -31,6 +31,9 @@ tracing = "0.1"
 tracing-subscriber = "0.3.9"
 tikv-jemalloc-sys = "0.4.3"
 parking_lot = {version = "0.12.0", features = ["arc_lock", "send_guard"]}
+hex-simd = "0.5.0"
```
[the remark this answers is not in the document]
Comment (Nugine) [reply Cargo.toml]: The latest commit formatted Cargo.toml by `cargo sort`.

File: src/device.rs (hunk 3 above)
Comment (Nugine) [reply src/device.rs]: Changed in the latest force-update.

**Comment by Nugine** [comment]
Close since the code has been published in another crate [rdma](https://github.com/Nugine/rdma).

**Review by rogercloud — APPROVED** [review]

**Pull request merged**